internal/controller/serverclaim_controller.go

```diff
@@ -251,6 +261,81 @@ func (r *ServerClaimReconciler) removeBootConfigRefFromServerAndPowerOff(ctx con
 	return nil
 }
 
+func (r *ServerClaimReconciler) getClaimableServer(ctx context.Context, log logr.Logger, claim *metalv1alpha1.ServerClaim) (*metalv1alpha1.Server, error) {
+	var (
+		server *metalv1alpha1.Server
+		err    error
+	)
+	switch {
+	case claim.Spec.ServerRef != nil:
+		server, err = r.serverByReference(ctx, log, claim)
+	case claim.Spec.ServerSelector != nil:
+		server, err = r.serverBySelector(ctx, log, claim)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if server == nil {
+		return nil, nil
+	}
+	// Check the server state and only proceed if it is Available or Reserved
+	if server.Status.State != metalv1alpha1.ServerStateAvailable && server.Status.State != metalv1alpha1.ServerStateReserved {
+		log.V(1).Info("Server not in a claimable state", "Server", server.Name, "ServerState", server.Status.State)
+		return nil, nil
+	}
+	return server, nil
+}
+
+func (r *ServerClaimReconciler) serverByReference(ctx context.Context, log logr.Logger, claim *metalv1alpha1.ServerClaim) (*metalv1alpha1.Server, error) {
+	server := &metalv1alpha1.Server{}
+	if err := r.Get(ctx, client.ObjectKey{Name: claim.Spec.ServerRef.Name}, server); err != nil {
+		return nil, err
+	}
+	if claimRef := server.Spec.ServerClaimRef; claimRef != nil && claimRef.UID != claim.UID {
+		log.V(1).Info("Server claim ref UID does not match claim", "Server", server.Name, "ClaimUID", claimRef.UID)
+		return nil, nil
+	}
+	if claim.Spec.ServerSelector == nil {
+		return server, nil
+	}
+	selector, err := metav1.LabelSelectorAsSelector(claim.Spec.ServerSelector)
```

Comment:
Why do we need to compare the selector labels in the case where we directly want to claim a `Server` via a reference?

Reply:
That's a good question. I'm not sure whether we should check labels or not. Let's say, it was done to initiate discussion. It's perfectly fine for me to remove this part.

Reply:
Resolved according to agreed approach https://github.com/ironcore-dev/metal-operator/pull/86#issuecomment-2233025341